File: backend/uchicago/lib/transcript_record.py
class TranscriptRecord(object):
    def __init__(self, section=None, grade=None, term=None):
        self.section = section
        self.grade = grade
        self.term = term

    @property
    def complete(self):
        return self.grade is not None

    def serialize(self):
        return {
            'term': self.term.id,
            'id': self.section.course.id,
            'section': self.section.id,
            'complete': self.complete,
            'quality': self.grade.quality if self.complete else None,
            'credit': self.grade.credit if self.complete else None,
            'grade': self.grade.grade if self.complete else None,
            'gpa': self.grade.gpa if self.complete else None
        }

File: frontend/src/app/transcript-record.ts
/**
 * A single transcript record.
 *
 * Do not add any class methods to this object, as it is deserialized via
 * typecasting.
 */
export class TranscriptRecord {
  course: string;
  complete: boolean;
  gpa: number;
  grade: string;
  quality: boolean;
  credit: boolean;
  section: string;
  term: string;

  static deserialize(data: any): TranscriptRecord {
    return <TranscriptRecord>data;
  }

  static compare(a: TranscriptRecord, b: TranscriptRecord): number {
    if (a.course < b.course) {
      return -1;
    }
    if (a.course > b.course) {
      return 1;
    }
    return 0;
  }
}
